azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```diff
@@ -0,0 +1,216 @@
+    this.missedSchedulesExecutor = this.missedSchedulesManagerEnabled
+        ? Executors.newFixedThreadPool(threadPoolSize, new DaemonThreadFactory("azk-missed-schedules-task-pool"))
+        : null;
+    if (this.missedSchedulesManagerEnabled) {
+      LOG.info("Missed Schedule Manager is ready to take tasks.");
+    } else {
+      LOG.info("Missed Schedule Manager is disabled.");
+    }
+  }
+
+  public boolean stop() throws InterruptedException {
+    if (this.missedSchedulesExecutor != null) {
+      this.missedSchedulesExecutor.shutdown();
+      if (!this.missedSchedulesExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
+        this.missedSchedulesExecutor.shutdownNow();
+      }
+    }
+    return true;
+  }
+
+  /**
+   * A MissedSchedule Task is a unit that store the critical information to execute a missedSchedule operation.
+   * A Task is capable to
+   * 1. notice user
+   * 2. execute back execution if enabled feature flag (TODO)
+   * when a missed schedule happens.
+   * */
+  public static class MissedSchedulesOperationTask implements Runnable {
+    private static final Logger LOG = LoggerFactory.getLogger(MissedSchedulesOperationTask.class);
+    private final Emailer emailer;
+    // emailRecipients to notify users the schedules missed
+    private final List<String> emailRecipients;
+    // the email body
+    private final String emailMessage;
+    // back execute action
+    // if user disabled the config, the action could be null.
+    private ExecuteFlowAction executeFlowAction;
+
+    public MissedSchedulesOperationTask(List<Long> missedScheduleTimes,
+        Emailer emailer,
+        List<String> emailRecipients,
+        String projectName,
+        String flowName) {
+      this.emailer = emailer;
+      this.emailRecipients = emailRecipients;
+      final String missScheduleTimestampToDate = formatTimestamps(missedScheduleTimes);
+      MessageFormat messageFormat = new MessageFormat("This is autogenerated email to notify to flow owners that"
+          + " the flow {0} in project {1} has scheduled the executions on {2} but failed to trigger on time.");
+      this.emailMessage = messageFormat.format(new Object[]{flowName, projectName, missScheduleTimestampToDate});
+      this.executeFlowAction = null;
+    }
+
+    public MissedSchedulesOperationTask(List<Long> missedScheduleTimes,
+        Emailer emailer,
+        List<String> emailRecipients,
+        ExecuteFlowAction executeFlowAction) {
+      this.emailer = emailer;
+      this.emailRecipients = emailRecipients;
+      this.executeFlowAction = executeFlowAction;
+      final String missScheduleTimestampToDate = formatTimestamps(missedScheduleTimes);
+      MessageFormat messageFormat = new MessageFormat("This is autogenerated email to notify to flow owners that"
+          + " the flow {0} in project {1} has scheduled the executions on {2} but failed to trigger on time. "
+          + "Back execution will start soon as you enabled the config.");
```

Comment:
Is this designed to handle the `backExecution` in some form, and what config is this referring to ?

Reply:
back execution config will be added as Schedule Parameter. User would need to update existing schedule to apply this config. It will be turned off by default. TriggerManager would aware of this config and passed into this Manager.
This will also come from later PR (change schedule servlet to support adding this config)